mlflow/langchain/__init__.py

```diff
@@ -314,13 +328,79 @@ def log_model(
 
 
 def _save_model(model, path):
-    model.save(path)
+    import langchain
+
+    model_data_path = os.path.join(path, _MODEL_DATA_FILE_NAME)
+    model_data_kwargs = {_MODEL_DATA_KEY: _MODEL_DATA_FILE_NAME}
+
+    if type(model) == langchain.chains.llm.LLMChain:
+        model.save(model_data_path)
+    elif type(model) == langchain.agents.agent.AgentExecutor:
+        if model.agent and model.agent.llm_chain:
+            model.agent.llm_chain.save(model_data_path)
+
+        if model.agent:
+            agent_data_path = os.path.join(path, _AGENT_DATA_FILE_NAME)
+            model.save_agent(agent_data_path)
+            model_data_kwargs[_AGENT_DATA_KEY] = _AGENT_DATA_FILE_NAME
+
+        if model.tools:
+            tools_data_path = os.path.join(path, _TOOLS_DATA_FILE_NAME)
+            with open(tools_data_path, "wb") as f:
+                cloudpickle.dump(model.tools, f)
+            model_data_kwargs[_TOOLS_DATA_KEY] = _TOOLS_DATA_FILE_NAME
+        else:
+            raise mlflow.MlflowException.invalid_parameter_value(
+                "For Initializing the AgentExecutor, tools must be provided."
+            )
+
+        key_to_ignore = ["llm_chain", "agent", "tools", "callback_manager"]
+        temp_dict = {}
+        for k, v in model.__dict__.items():
+            if k not in key_to_ignore:
+                temp_dict[k] = v
+
+        agent_primitive_path = os.path.join(path, _AGENT_PRIMITIVES_FILE_NAME)
+        with open(agent_primitive_path, "w") as config_file:
+            json.dump(temp_dict, config_file, indent=4)
+
+        model_data_kwargs[_AGENT_PRIMITIVES_DATA_KEY] = _AGENT_PRIMITIVES_FILE_NAME
+    else:
+        raise mlflow.MlflowException.invalid_parameter_value(
+            "MLflow langchain flavor only supports logging langchain.chains.llm.LLMChain and "
+            + f"langchain.agents.agent.AgentExecutor instances, found {type(lc_model)}"
+        )
 
+    return model_data_kwargs
 
-def _load_model(path):
-    from langchain.chains.loading import load_chain
 
-    model = load_chain(path)
+def _load_model(path, agent_path=None, tools_path=None, agent_primitive_path=None):
+    model = None
+    if agent_path is None and tools_path is None:
+        from langchain.chains.loading import load_chain
+
+        model = load_chain(path)
+    else:
+        from langchain.chains.loading import load_chain
+        from langchain.agents import initialize_agent
+
+        llm = load_chain(path)
+        tools = []
+        args = {}
+
+        if os.path.exists(tools_path):
+            with open(tools_path, "rb") as f:
+                tools = cloudpickle.load(f)
+        else:
+            raise mlflow.MlflowException.invalid_parameter_value(
+                "For Initializing the AgentExecutor, tools must be provided."
```

Comment:
Can we be more specific about the source of the error, which is that the expected tools file is missing? This would be an internal error, right?